### PR TITLE
Package updates (Gulp v4 + Jest v24 upgrades)

### PR DESCRIPTION
**Pull request by colinrotherham**
As discussed in https://github.com/alphagov/govuk-frontend/issues/1273 there were various missing `peerDependencies` and 290 npm vulnerabilities found after install.

It did involve a few breaking changes:

**1. Gulp v4**
Lots of minor API changes

**2. Jest v24**
Can't use `__BROWSER__` or any globals so moved to `jest-puppeteer`

You get a nice config file to configure puppeteer `./jest-puppeteer.config.js` to handle starting the application, and means we can remove all the `./lib/puppeteer` boilerplate.

### CODE REVIEW

**Comment by 36degrees** [comment]
Hey @colinrotherham,

Thanks, this is awesome work. There's a couple of things that need looking at…

It looks like the tests that get run after building dist or the package are failing on this branch.

```
$ npm run build:dist

> govuk-frontend-repository@ build:dist /Users/oliverbyford/Code/govuk-frontend
> node bin/check-nvmrc.js && gulp build:dist --destination 'dist' && npm run test:build:dist

[11:22:19] Using gulpfile ~/Code/govuk-frontend/gulpfile.js
[11:22:19] Starting 'build:dist'...
[11:22:19] Starting 'clean'...
[11:22:19] Finished 'clean' after 41 ms
[11:22:19] Starting 'copy-assets'...
[11:22:19] Starting 'styles'...
[11:22:19] Starting 'scss:lint'...
[11:22:20] Finished 'scss:lint' after 1.14 s
[11:22:20] Starting 'scss:compile'...
[11:22:22] Finished 'scss:compile' after 2.45 s
[11:22:22] Finished 'styles' after 3.59 s
[11:22:22] Starting 'scripts'...
[11:22:22] Starting 'js:compile'...
[11:22:23] Finished 'js:compile' after 738 ms
[11:22:23] Finished 'scripts' after 738 ms
[11:22:23] Finished 'copy-assets' after 4.33 s
[11:22:23] Starting 'copy:assets'...
[11:22:23] Finished 'copy:assets' after 43 ms
[11:22:23] Starting 'update-assets-version'...
[11:22:23] Finished 'update-assets-version' after 10 ms
[11:22:23] Finished 'build:dist' after 4.43 s

> govuk-frontend-repository@ test:build:dist /Users/oliverbyford/Code/govuk-frontend
> jest tasks/gulp/__tests__/after-build-dist.test.js

No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/oliverbyford/Code/govuk-frontend
  333 files checked.
  testMatch: **/__tests__/**/*.[jt]s?(x), **/?(*.)+(spec|test).[tj]s?(x) - 181 matches
  testPathIgnorePatterns: /node_modules/, /public/, after-* - 252 matches
  testRegex:  - 0 matches
Pattern: tasks/gulp/__tests__/after-build-dist.test.js - 0 matches
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! govuk-frontend-repository@ test:build:dist: `jest tasks/gulp/__tests__/after-build-dist.test.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the govuk-frontend-repository@ test:build:dist script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

```
$ npm run build:package

> govuk-frontend-repository@ build:package /Users/oliverbyford/Code/govuk-frontend
> node bin/check-nvmrc.js && gulp build:package --destination 'package' && npm run test:build:package

[11:21:14] Using gulpfile ~/Code/govuk-frontend/gulpfile.js
[11:21:14] Starting 'build:package'...
[11:21:14] Starting 'clean'...
[11:21:14] Finished 'clean' after 120 ms
[11:21:14] Starting 'copy-files'...
[11:21:15] Finished 'copy-files' after 1.1 s
[11:21:15] Starting 'js:compile'...
[11:21:16] Finished 'js:compile' after 1.24 s
[11:21:16] Finished 'build:package' after 2.46 s

> govuk-frontend-repository@ test:build:package /Users/oliverbyford/Code/govuk-frontend
> jest tasks/gulp/__tests__/after-build-package.test.js

No tests found, exiting with code 1
Run with `--passWithNoTests` to exit with code 0
In /Users/oliverbyford/Code/govuk-frontend
  333 files checked.
  testMatch: **/__tests__/**/*.[jt]s?(x), **/?(*.)+(spec|test).[tj]s?(x) - 181 matches
  testPathIgnorePatterns: /node_modules/, /public/, after-* - 252 matches
  testRegex:  - 0 matches
Pattern: tasks/gulp/__tests__/after-build-package.test.js - 0 matches
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! govuk-frontend-repository@ test:build:package: `jest tasks/gulp/__tests__/after-build-package.test.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the govuk-frontend-repository@ test:build:package script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

The tests are also being copied to `package` when `npm run build:package` is run, which [shouldn't be happening](https://github.com/alphagov/govuk-frontend/blob/master/tasks/gulp/copy-to-destination.js#L22).

File: package.json [hunk 1]
```diff
@@ -16,20 +16,22 @@
     "release-to-branch": "node bin/check-nvmrc.js && ./bin/release-to-branch.sh",
     "build:package": "node bin/check-nvmrc.js && gulp build:package --destination 'package' && npm run test:build:package",
     "build:dist": "node bin/check-nvmrc.js && gulp build:dist --destination 'dist' && npm run test:build:dist",
-    "test": "standard && gulp test && gulp copy-assets && jest --testPathIgnorePatterns='after-*'",
+    "test": "npm ls --depth=0 && standard && gulp test && gulp copy-assets && jest",
```
Comment (36degrees) [inline package.json]: Could we either add a comment or break this into its own named task to make it clear what this is doing? I initially thought that this had been left in from debugging, as I wouldn't expect the exit code of `npm ls` to be important.

**Comment by colinrotherham** [comment]
Thanks @36degrees, bound to be a few things like this! So many little API changes in Jest, will be good to get it up to date again.

I'll take a look again.

File: package.json (hunk 1 above)
Comment (colinrotherham) [reply package.json]: Yes will do.

It checks the dependency tree for missing packages (you had a handful).

Can't put comments in JSON so I'll make the script name more descriptive.

**Comment by colinrotherham** [comment]
A side effect Gulp v4 globbing in `gulp.src()` is that negation must be done last, not first.

Going to fix this now.

### Negating first
Tests are included

```js
return gulp.src([
  '!' + configPaths.src + '**/*.test.js',
  srcFiles
])
```

```bash
[13:02:57] Using gulpfile ~/Sites/GDS/govuk-frontend/gulpfile.js
[13:02:57] Starting 'js:compile'...
[13:02:57] gulp-debug: src/all.js
[13:02:57] gulp-debug: src/all.test.js
[13:02:57] gulp-debug: src/common.js
[13:02:57] gulp-debug: src/template.test.js
[13:02:57] gulp-debug: src/core/core.test.js
[13:02:57] gulp-debug: src/components/all.test.js
[13:02:57] gulp-debug: src/helpers/colour.test.js
[13:02:57] gulp-debug: src/helpers/grid.test.js
[13:02:57] gulp-debug: src/helpers/helpers.test.js
[13:02:57] gulp-debug: src/helpers/media-queries.test.js
[13:02:57] gulp-debug: src/helpers/spacing.test.js
[13:02:57] gulp-debug: src/helpers/typography.test.js
[13:02:57] gulp-debug: src/objects/objects.test.js
[13:02:57] gulp-debug: src/overrides/overrides.test.js
[13:02:57] gulp-debug: src/settings/colours.test.js
[13:02:57] gulp-debug: src/settings/settings.test.js
[13:02:57] gulp-debug: src/tools/compatibility.test.js
[13:02:57] gulp-debug: src/tools/exports.test.js
[13:02:57] gulp-debug: src/tools/font-url.test.js
[13:02:57] gulp-debug: src/tools/iff.test.js
[13:02:57] gulp-debug: src/tools/image-url.test.js
[13:02:57] gulp-debug: src/tools/tools.test.js
[13:02:57] gulp-debug: src/components/accordion/accordion.js
[13:02:57] gulp-debug: src/components/accordion/accordion.test.js
[13:02:57] gulp-debug: src/components/accordion/template.test.js
[13:02:57] gulp-debug: src/components/back-link/template.test.js
[13:02:57] gulp-debug: src/components/breadcrumbs/template.test.js
[13:02:57] gulp-debug: src/components/button/button.js
[13:02:58] gulp-debug: src/components/button/button.test.js
[13:02:58] gulp-debug: src/components/button/template.test.js
[13:02:58] gulp-debug: src/components/character-count/character-count.js
[13:02:58] gulp-debug: src/components/character-count/character-count.test.js
[13:02:58] gulp-debug: src/components/character-count/template.test.js
[13:02:58] gulp-debug: src/components/checkboxes/checkboxes.js
[13:02:58] gulp-debug: src/components/checkboxes/checkboxes.test.js
[13:02:58] gulp-debug: src/components/checkboxes/template.test.js
[13:02:58] gulp-debug: src/components/date-input/template.test.js
[13:02:58] gulp-debug: src/components/error-message/template.test.js
[13:02:58] gulp-debug: src/components/details/details.js
[13:02:58] gulp-debug: src/components/details/details.test.js
[13:02:58] gulp-debug: src/components/details/template.test.js
[13:02:58] gulp-debug: src/components/error-summary/error-summary.js
[13:02:58] gulp-debug: src/components/error-summary/error-summary.test.js
[13:02:58] gulp-debug: src/components/error-summary/template.test.js
[13:02:59] gulp-debug: src/components/fieldset/template.test.js
[13:02:59] gulp-debug: src/components/file-upload/template.test.js
[13:02:59] gulp-debug: src/components/footer/template.test.js
[13:02:59] gulp-debug: src/components/header/header.js
[13:02:59] gulp-debug: src/components/header/header.test.js
[13:02:59] gulp-debug: src/components/header/template.test.js
[13:02:59] gulp-debug: src/components/hint/template.test.js
[13:02:59] gulp-debug: src/components/input/template.test.js
[13:02:59] gulp-debug: src/components/inset-text/template.test.js
[13:02:59] gulp-debug: src/components/label/template.test.js
[13:02:59] gulp-debug: src/components/panel/template.test.js
[13:02:59] gulp-debug: src/components/phase-banner/template.test.js
[13:02:59] gulp-debug: src/components/radios/radios.js
[13:02:59] gulp-debug: src/components/radios/radios.test.js
[13:02:59] gulp-debug: src/components/radios/template.test.js
[13:02:59] gulp-debug: src/components/select/template.test.js
[13:02:59] gulp-debug: src/components/skip-link/template.test.js
[13:02:59] gulp-debug: src/components/summary-list/template.test.js
[13:02:59] gulp-debug: src/components/table/template.test.js
[13:02:59] gulp-debug: src/components/tabs/tabs.js
[13:02:59] gulp-debug: src/components/tabs/tabs.test.js
[13:02:59] gulp-debug: src/components/tabs/template.test.js
[13:02:59] gulp-debug: src/components/tag/template.test.js
[13:03:00] gulp-debug: src/components/textarea/template.test.js
[13:03:00] gulp-debug: src/components/warning-text/template.test.js
[13:03:00] gulp-debug: src/vendor/polyfills/DOMTokenList.js
[13:03:00] gulp-debug: src/vendor/polyfills/Document.js
[13:03:00] gulp-debug: src/vendor/polyfills/Element.js
[13:03:00] gulp-debug: src/vendor/polyfills/Event.js
[13:03:00] gulp-debug: src/vendor/polyfills/Window.js
[13:03:00] gulp-debug: src/vendor/polyfills/Object/defineProperty.js
[13:03:00] gulp-debug: src/vendor/polyfills/Element/prototype/classList.js
[13:03:00] gulp-debug: src/vendor/polyfills/Element/prototype/closest.js
[13:03:00] gulp-debug: src/vendor/polyfills/Element/prototype/matches.js
[13:03:00] gulp-debug: src/vendor/polyfills/Function/prototype/bind.js
[13:03:00] gulp-debug: 79 items
[13:03:00] Finished 'js:compile' after 3.37 s
```

### Negating last
Tests aren't included

```js
return gulp.src([
  srcFiles,
  '!' + configPaths.src + '**/*.test.js'
])
```

```bash
[13:00:39] gulp-debug: src/all.js
[13:00:39] gulp-debug: src/common.js
[13:00:39] gulp-debug: src/components/accordion/accordion.js
[13:00:40] gulp-debug: src/components/button/button.js
[13:00:40] gulp-debug: src/components/character-count/character-count.js
[13:00:40] gulp-debug: src/components/checkboxes/checkboxes.js
[13:00:40] gulp-debug: src/components/details/details.js
[13:00:40] gulp-debug: src/components/error-summary/error-summary.js
[13:00:40] gulp-debug: src/components/header/header.js
[13:00:40] gulp-debug: src/components/radios/radios.js
[13:00:40] gulp-debug: src/components/tabs/tabs.js
[13:00:40] gulp-debug: src/vendor/polyfills/DOMTokenList.js
[13:00:40] gulp-debug: src/vendor/polyfills/Document.js
[13:00:40] gulp-debug: src/vendor/polyfills/Element.js
[13:00:40] gulp-debug: src/vendor/polyfills/Event.js
[13:00:40] gulp-debug: src/vendor/polyfills/Window.js
[13:00:40] gulp-debug: src/vendor/polyfills/Object/defineProperty.js
[13:00:40] gulp-debug: src/vendor/polyfills/Element/prototype/classList.js
[13:00:40] gulp-debug: src/vendor/polyfills/Element/prototype/closest.js
[13:00:40] gulp-debug: src/vendor/polyfills/Element/prototype/matches.js
[13:00:41] gulp-debug: src/vendor/polyfills/Function/prototype/bind.js
[13:00:41] gulp-debug: 21 items
[13:00:41] Finished 'js:compile' after 1.33 s
```

File: package.json (hunk 1 above)
Comment (colinrotherham) [reply package.json]: @36degrees I've resolved this with a new `npm run test:dependencies` task that's run first before any tests are started.

**Comment by colinrotherham** [comment]
@36degrees The two build task failures were caused by `testPathIgnorePatterns` being moved into the Jest config at the bottom of **package.json**. It's now back in the main _test_ npm script.

All resolved now 👍

**Comment by 36degrees** [comment]
Result of running `git diff` after `npm run build:dist` and `npm run build:package`:

```diff
diff --git a/dist/govuk-frontend-2.10.0.min.css b/dist/govuk-frontend-2.10.0.min.css
index b405e1f0..2f083055 100644
--- a/dist/govuk-frontend-2.10.0.min.css
+++ b/dist/govuk-frontend-2.10.0.min.css
@@ -1608,7 +1608,6 @@
 }
 
 @media only screen and (-webkit-min-device-pixel-ratio:2),
-only screen and (min--moz-device-pixel-ratio:2),
 only screen and (min-device-pixel-ratio:2),
 only screen and (min-resolution:2dppx),
 only screen and (min-resolution:192dpi) {
@@ -2999,7 +2998,6 @@ only screen and (min-resolution:192dpi) {
 }
 
 @media only screen and (-webkit-min-device-pixel-ratio:2),
-only screen and (min--moz-device-pixel-ratio:2),
 only screen and (min-device-pixel-ratio:2),
 only screen and (min-resolution:2dppx),
 only screen and (min-resolution:192dpi) {
@@ -3091,17 +3089,20 @@ only screen and (min-resolution:192dpi) {
     padding: 0;
     list-style: none;
     -webkit-column-gap: 30px;
+    -moz-column-gap: 30px;
     column-gap: 30px
 }
 
 @media (min-width:48.0625em) {
     .govuk-footer__list--columns-2 {
         -webkit-column-count: 2;
+        -moz-column-count: 2;
         column-count: 2
     }
 
     .govuk-footer__list--columns-3 {
         -webkit-column-count: 3;
+        -moz-column-count: 3;
         column-count: 3
     }
 }
diff --git a/dist/govuk-frontend-ie8-2.10.0.min.css b/dist/govuk-frontend-ie8-2.10.0.min.css
index abb63c90..26c62254 100644
--- a/dist/govuk-frontend-ie8-2.10.0.min.css
+++ b/dist/govuk-frontend-ie8-2.10.0.min.css
@@ -1892,16 +1892,19 @@
     padding: 0;
     list-style: none;
     -webkit-column-gap: 30px;
+    -moz-column-gap: 30px;
     column-gap: 30px
 }
 
 .govuk-footer__list--columns-2 {
     -webkit-column-count: 2;
+    -moz-column-count: 2;
     column-count: 2
 }
 
 .govuk-footer__list--columns-3 {
     -webkit-column-count: 3;
+    -moz-column-count: 3;
     column-count: 3
 }
 
diff --git a/package/components/footer/_footer.scss b/package/components/footer/_footer.scss
index ec004225..9ee8181b 100644
--- a/package/components/footer/_footer.scss
+++ b/package/components/footer/_footer.scss
@@ -208,17 +208,20 @@
     padding: 0;
     list-style: none;
     -webkit-column-gap: $govuk-gutter;
+       -moz-column-gap: $govuk-gutter;
             column-gap: $govuk-gutter; // Support: Columns
   }
 
   @include mq ($from: desktop) {
     .govuk-footer__list--columns-2 {
       -webkit-column-count: 2;
+         -moz-column-count: 2;
               column-count: 2; // Support: Columns
     }
 
     .govuk-footer__list--columns-3 {
       -webkit-column-count: 3;
+         -moz-column-count: 3;
               column-count: 3; // Support: Columns
     }
   }
diff --git a/package/helpers/_device-pixels.scss b/package/helpers/_device-pixels.scss
index e8202b55..3b860025 100644
--- a/package/helpers/_device-pixels.scss
+++ b/package/helpers/_device-pixels.scss
@@ -29,7 +29,6 @@
 
 @mixin govuk-device-pixel-ratio($ratio: 2) {
   @media only screen and (-webkit-min-device-pixel-ratio: $ratio),
-    only screen and (min--moz-device-pixel-ratio: $ratio),
     only screen and (     min-device-pixel-ratio: $ratio),
     only screen and (             min-resolution: #{($ratio*96)}dpi),
     only screen and (             min-resolution: #{$ratio}dppx) {
```

(There were also some changes to `package-lock.json` after running `npm install` – mainly some `sha1` hashes changing to `sha256` and some packages no longer being marked as optional.)

**Comment by 36degrees** [comment]
I _think_ the watch task is no longer working correctly – if I'm running the review app and make changes (either to the app or to `/src` then the styles / scripts don't appear to be getting compiled and nothing changes.

**Comment by colinrotherham** [comment]
Thanks @36degrees.

Not sure how to resolve the diff comments above?

Because we've made package updates, autoprefixer output has changed:
https://github.com/alphagov/govuk-frontend/pull/1277/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R26

Should these be included in a commit?

Will look into the Gulp v4 watch changes now.

**Comment by 36degrees** [comment]
Sorry, my fault for being terse – was just documenting that as I was testing it, I don't believe it's an issue that the output has changed, as you say it's related to the autoprefixer version.

**Comment by colinrotherham** [comment]
@36degrees Fab. It looks like the changes get picked up and committed at the next release so we're all good?

Should the watch task have started as part of `npm start` or do you run other tasks during development?

I'll take a look.

**Comment by 36degrees** [comment]
> Fab. It looks like the changes get picked up and committed at the next release so we're all good?

Yep, we run the build tasks whenever we do a release – I'm just running the build tasks now to make sure they work as expected :)

> Should the watch task have started as part of npm start or do you run other tasks during development?

Normally running `npm start` is enough… I _think_ [the `serve` task used to depend on the `watch` task](https://github.com/alphagov/govuk-frontend/pull/1277/files#diff-b9e12334e9eafd8341a6107dd98510c9L83), but that's [no longer the case](https://github.com/alphagov/govuk-frontend/pull/1277/files#diff-b9e12334e9eafd8341a6107dd98510c9R63)?

**Comment by colinrotherham** [comment]
That'll be it then, well spotted. That feature was removed in Gulp v4 so it'll need adding back in the list of dev tasks.

**Comment by colinrotherham** [comment]
@36degrees The `watch` task is now back in as part of `serve`:

```js
gulp.task('serve', gulp.parallel(
  'watch',
  'nodemon'
))
```

I've moved **nodemon** to a named gulp task in `'./tasks/gulp'` (rather than an anonymous function), as the Terminal output from Gulp v4 wasn't as obvious:

#### Before
```bash
[17:49:55] Starting 'serve'...
[17:49:55] Starting 'watch'...
[17:49:55] Starting '<anonymous>'...
```

#### After
```bash
[17:43:02] Starting 'serve'...
[17:43:02] Starting 'watch'...
[17:43:02] Starting 'nodemon'...
```

File: tasks/gulp/compile-assets.js [hunk 2]
```diff
@@ -143,8 +143,8 @@ gulp.task('js:compile', () => {
   // for dist/ folder we only want compiled 'all.js' file
   let srcFiles = isDist ? configPaths.src + 'all.js' : configPaths.src + '**/*.js'
   return gulp.src([
-    '!' + configPaths.src + '**/*.test.js',
-    srcFiles
+    srcFiles,
+    '!' + configPaths.src + '**/*.test.js'
```
Comment (hannalaakso) [inline tasks/gulp/compile-assets.js]: Non-blocking: maybe we could have a comment here to say why the order of these is important (I saw that this was already discussed in the comments on the PR)

Comment (colinrotherham) [reply tasks/gulp/compile-assets.js]: Ah sorry @hannalaakso  it's been merged already.

If it helps, the order was already correct in these two:

```
./tasks/gulp/copy-to-destination.js
./tasks/gulp/sassdoc.js
```

For reference, latest docs are here:
https://github.com/gulpjs/glob-stream#globbing--negation

Seems to be performance related to now glob based on the specified order.

**Pull request merged**